src/main/scala/com/box/castle/router/kafkadispatcher/cache/CacheWithData.scala

```diff
@@ -23,112 +25,173 @@ private[cache] class CacheWithData private(val data: LinkedHashMap[Long, CastleM
 
     // First we try to get the CastleMessageBatch directly from our data map associated with the offset
-    data.get(offset).orElse {
+    val firstBatch = data.get(offset).orElse {
       // Otherwise we will look inside each CastleMessageBatch and try to "slice" out the messages
       // from within the batch, but we can only do this if the offset falls within our range of the minimum and
       // maximum offset we have across all the CastleMessageBatches within our data map
       if (offset >= minOffset && offset <= maxOffset)
-        // We iterate over all of the castle message batches we have in data and attempt to
-        // create a batch from the given offset if it happens to fall inside that batch
-        // createBatchFromOffset returns Some(CastleMessageBatch) if it is successful and None otherwise.
+      // We iterate over all of the castle message batches we have in data and attempt to
+      // create a batch from the given offset if it happens to fall inside that batch
+      // createBatchFromOffset returns Some(CastleMessageBatch) if it is successful and None otherwise.
         findOption(data.values)(castleMessageBatch => castleMessageBatch.createBatchFromOffset(offset))
       else
         None
     }
+
+    firstBatch match {
+      case Some(batch) =>
+        // Cache Hit
+        if (bufferSize > 0 && bufferSize < batch.sizeInBytes) {
+          // Buffer too small for this batch
+          None
+        }
+        else {
+          // Try to fetch more contiguous batches form cache
+          val messageBatches = getMoreRecursive(Vector(batch), bufferSize - batch.sizeInBytes)
+          if (messageBatches.size > 1)
+            Some(CastleMessageBatch(messageBatches)) // Concatenate batches
+          else
+            Some(messageBatches.head)
+        }
+      case None =>
+        // Cache miss
+        None
+    }
   }
 
   /**
-   * findOption iterates through an Iterable of A and applies function f to each value of type A
-   * which returns back an Option[B]
-   * If Option[B] is not empty findOption will return that as the result,
-   * Otherwise findOption continues going through the remaining values.
-   * If findOption exhausts all values without encounering a Some value, it returns None
-   */
+    * Recursive function to keep fetching more batches from the cache with the nextOffset of the previous batch
+    * till a cache miss happens or we exhaust the buffer capacity. If a whole batch does not fit, it maybe sliced to
+    * fit partially.
+    *
+    * @param currentList
+    * @param remainingBufferSize
+    * @return
+    */
   @tailrec
-  private def findOption[A,B](values: Iterable[A])(f: A => Option[B]): Option[B] =
-    values.headOption match {
-      case Some(h) =>
-        f(h) match {
-          case Some(v) => Some(v)
-          case None => findOption(values.tail)(f)
+  private def getMoreRecursive(currentList: IndexedSeq[CastleMessageBatch],
```

Comment:
why do you need an IndexedSeq here? It seems a Seq would be enough.

Reply:
In line 136 I am doing an append at the end of currentList which I think is more performant with IndexedSeq than Seq.